### PR TITLE
[Enhancement](restore) restore table or partition with the same replication num as before the backup 

### LINKED ISSUE #11941: [Enhancement] (restore) restore table or partition with the same replication_num as before the backup
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

right now, if i restore a table, i must set "replication_num"="1" or other replication num or default 3 replication, and i will get a table with all parititions which have same replication_num.

if a table have different replication num, if backup and restore it, i can not get a table which every partition with the same replication num as before the backup.

for example:
every partition of a table do not have same  replication_num:
MySQL [tmp]> show partitions from part_p;
+-------------+---------------+----------------+---------------------+--------+--------------+------------------------------------------------------------------------+-----------------+---------+----------------+---------------+---------------------+---------------------+--------------------------+----------+------------+-------------------------+
| PartitionId | PartitionName | VisibleVersion | VisibleVersionTime  | State  | PartitionKey | Range                                                                  | DistributionKey | Buckets | ReplicationNum | StorageMedium | CooldownTime        | RemoteStoragePolicy | LastConsistencyCheckTime | DataSize | IsInMemory | ReplicaAllocation       |
+-------------+---------------+----------------+---------------------+--------+--------------+------------------------------------------------------------------------+-----------------+---------+----------------+---------------+---------------------+---------------------+--------------------------+----------+------------+-------------------------+
| 16004       | p1            | 2              | 2022-08-20 17:14:51 | NORMAL | p_partkey    | [types: [INT]; keys: [-2147483648]; ..types: [INT]; keys: [1013157]; ) | p_partkey       | 3       | 1              | HDD           | 9999-12-31 23:59:59 |                     | NULL                     | 8.634 MB | false      | tag.location.default: 1 |
| 16005       | p2            | 2              | 2022-08-20 17:14:51 | NORMAL | p_partkey    | [types: [INT]; keys: [1013157]; ..types: [INT]; keys: [1113157]; )     | p_partkey       | 3       | 2              | HDD           | 9999-12-31 23:59:59 |                     | NULL                     | 2.175 MB | false      | tag.location.default: 2 |
| 16006       | p3            | 2              | 2022-08-20 17:14:51 | NORMAL | p_partkey    | [types: [INT]; keys: [1113157]; ..types: [INT]; keys: [1213157]; )     | p_partkey       | 3       | 3              | HDD           | 9999-12-31 23:59:59 |                     | NULL                     | 3.265 MB | false      | tag.location.default: 3 |
| 16007       | p4            | 2              | 2022-08-20 17:14:51 | NORMAL | p_partkey    | [types: [INT]; keys: [1213157]; ..types: [INT]; keys: [2147483647]; )  | p_partkey       | 3       | 1              | HDD           | 9999-12-31 23:59:59 |                     | NULL                     | 1.810 MB | false      | tag.location.default: 1 |
+-------------+---------------+----------------+---------------------+--------+--------------+------------------------------------------------------------------------+-----------------+---------+----------------+---------------+---------------------+---------------------+--------------------------+----------+------------+-------------------------+
4 rows in set (0.00 sec)

this table have some partitions which have different replication num.

i backup this table and restore it, i can not get same table with same partitions with the same replication num as before the backup.

### Solution

add restore new property 'reserve_replica', which means you can get a table with same partitions with the same replication num as before the backup.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nextdreamblue**
add restore new property 'reserve_replica', which means you can
get a table with same partitions with the same replication num
as before the backup.

# Proposed changes

Issue Number: close #11941

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
2. Has unit tests been added:
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
4. Does it need to update dependencies:
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/zh-CN/docs/sql-manual/sql-reference/Data-Definition-Statements/Backup-and-Restore/RESTORE.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ PROPERTIES ("key"="value", ...);
 - PROPERTIES 目前支持以下属性：
   -  "backup_timestamp" = "2018-05-04-16-45-08"：指定了恢复对应备份的哪个时间版本，必填。该信息可以通过 `SHOW SNAPSHOT ON repo;` 语句获得。
   - "replication_num" = "3"：指定恢复的表或分区的副本数。默认为3。若恢复已存在的表或分区，则副本数必须和已存在表或分区的副本数相同。同时，必须有足够的 host 容纳多个副本。
+  - "reserve_replica" = "true":默认为false。当该属性为true时，会忽略'replication_num'属性，恢复的表或分区的副本数将与备份之前一样。支持多个表或表内多个分区有不同的副本数。
```
Comment (stalary) [inline docs/zh-CN/docs/sql-manual/sql-reference/Data-Definition-Statements/Backup-and-Restore/RESTORE.md:59]:
```suggestion
  - "reserve_replica" = "true"：默认为 false。当该属性为 true 时，会忽略 replication_num 属性，恢复的表或分区的副本数将与备份之前一样。支持多个表或表内多个分区有不同的副本数。
```

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 2]
```diff
@@ -162,14 +166,17 @@ public RestoreJob() {
...
+            long repoId) {
         super(JobType.RESTORE, label, dbId, dbName, timeoutMs, env, repoId);
         this.backupTimestamp = backupTs;
         this.jobInfo = jobInfo;
         this.allowLoad = allowLoad;
         this.replicaAlloc = replicaAlloc;
         this.state = RestoreJobState.PENDING;
         this.metaVersion = metaVersion;
+        this.reserveReplica = reserveReplica;
+        properties.put(PROP_RESERVE_REPLICA, String.valueOf(reserveReplica));
```
Comment (stalary) [inline fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:179]: Why assign a value here, which looks different from the rest of the parameters?

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 3]
```diff
@@ -1858,6 +1893,7 @@ public void readFields(DataInput in) throws IOException {
             String value = Text.readString(in);
             properties.put(key, value);
         }
+        reserveReplica = Boolean.parseBoolean(properties.get(PROP_RESERVE_REPLICA));
```
Comment (stalary) [inline fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:1896]: What does this do

**Review by stalary — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 3 above)
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:1896]: 主要是为了兼容老的元数据，这个新的属性通过properties这个已有的变量去保存，可以很好地兼容之前的恢复任务元数据。如果不保存，会有丢失这个属性信息，所以折中了下，使用properties这个目前看来其实没有使用的变量。

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 2 above)
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:179]: 理由同下

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 3 above)
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:1896]: 如果不这样保存，而是使用单独一个值去write和read，那么老的没有该值的恢复任务在加载的时候，直接会报错，元数据读取失败。

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 2 above)
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:179]: 把该值放到properties，只是为了通过properties来持久化到元数据，否则元数据会有前后兼容问题

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 3 above)
Comment (stalary) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:1896]: In case there is no value, set `false`?

Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:1896]: 对，默认值是false。兼容老的restore任务信息

File: fe/fe-core/src/main/java/org/apache/doris/analysis/RestoreStmt.java [hunk 4]
```diff
@@ -106,6 +112,18 @@ public void analyzeProperties() throws AnalysisException {
...
         }
+        // reserve replica
+        if (copiedProperties.containsKey(PROP_RESERVE_REPLICA)) {
+            if (copiedProperties.get(PROP_RESERVE_REPLICA).equalsIgnoreCase("true")) {
+                reserveReplica = true;
+            } else if (copiedProperties.get(PROP_RESERVE_REPLICA).equalsIgnoreCase("false")) {
+                reserveReplica = false;
+            } else {
+                ErrorReport.reportAnalysisException(ErrorCode.ERR_COMMON_ERROR,
+                        "Invalid reserve replica value: " + copiedProperties.get(PROP_RESERVE_REPLICA));
```
Comment (cambyzju) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/RestoreStmt.java:123]:
```suggestion
                        "Invalid reserve_replica value: " + copiedProperties.get(PROP_RESERVE_REPLICA));
```

**Review by lide-reed — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java (hunk 2 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:179]: Looks like the `properties` is indeed build for this purpose. I think it is ok to do so.

**Review by stalary — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java [hunk 5]
```diff
@@ -927,9 +948,18 @@ private void checkAndRestoreResources() {
     }
 
     private boolean genFileMappingWhenBackupReplicasEqual(PartitionInfo localPartInfo, Partition localPartition,
-            Table localTbl, BackupPartitionInfo backupPartInfo, String partitionName, BackupOlapTableInfo tblInfo) {
-        short restoreReplicaNum = replicaAlloc.getTotalReplicaNum();
-        short localReplicaNum = localPartInfo.getReplicaAllocation(localPartition.getId()).getTotalReplicaNum();
+            Table localTbl, BackupPartitionInfo backupPartInfo, String partitionName, BackupOlapTableInfo tblInfo,
```
[the remark this answers is not in the document]
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/backup/RestoreJob.java:951]: fix code done

**Review by lide-reed — APPROVED** [review]
LGTM

**Pull request merged**